### PR TITLE
♿  [Story a11y] Improve tabbing on social-share and bookend

### PR DESCRIPTION
**Pull request by mszylkowski**
Part of #31747
Mentioned in #32511

When tabbing through the story, landing on the share icons or the bookend links will scroll the viewport outside the story area.

Changes:
- When showing or hiding the share menu, toggle the share icons to tabbable.
- Toggle the links on the bookend to tabbable when bookend is open
- Bring back focus state:
image

@ampproject/wg-stories

### CODE REVIEW

**Comment by processprocess** [comment]
An alternate approach:
putting `visibility: hidden` on the parent when it's not open might prevent the elements from being tabbable. 

It may have side effects depending on which node is being transformed.
I'm also not sure how that works with shadow dom. It might need to be on a parent that's in the same shadow dom.

Do you think there's value in trying that? It might be less code.

**Comment by mszylkowski** [comment]
@processprocess Absolutely, let me try that. I suspect that the animations would break since we're hiding the element when not visible

**Comment by mszylkowski** [comment]
@processprocess seems like the shadow dom does interfere but it was worth the effort to get it working through the shadow dom since the code is much clearer this way, PTAL

**Comment by processprocess** [comment]
This looks great to me.

I noticed "Get link" gets skipped when tabbing. It looks like screen readers don't recognize it. I think that needs to be fixed.
It's somewhat related to this PR. Do you think it makes sense to do that work in this PR or split it into a new one?

Screen Shot 2021-02-25 at 10 32 12 AM

File: extensions/amp-story/1.0/amp-story-bookend.css [hunk 1]
```diff
@@ -26,6 +26,11 @@
   color: #fff !important;
   text-align: start !important;
   z-index: 100001 !important;
+  visibility: hidden !important;
+}
+
+.i-amphtml-story-bookend[active] {
+  visibility: initial !important;
 }
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-bookend.css:34]: I wonder if this needs a comment that it's just for a11y.
Other animations happen so it might be good to note that this is unrelated to visual changes.

**Review by gmajoulet — COMMENTED** [review]
We already set `aria-hidden=bool` from the draggable-drawer class, that's supposed to take care of this

https://github.com/ampproject/amphtml/blob/476d99c927db659ffcf2ccde286552dbca7aeaef/extensions/amp-story/1.0/amp-story-draggable-drawer.js#L505

Why is this not working? Let's fix it instead of introducing yet another way to do the same thing

**Comment by mszylkowski** [comment]
> I noticed "Get link" gets skipped when tabbing. It looks like screen readers don't recognize it. I think that needs to be fixed.
It's somewhat related to this PR. Do you think it makes sense to do that work in this PR or split it into a new one?

@processprocess  I'll check if the "copy to clipboard" can be triggered from the keyboard, and if it's a couple of lines of code we can do it here.

> We already set aria-hidden=bool from the draggable-drawer class, that's supposed to take care of this

@gmajoulet https://web.dev/aria-hidden-focus/ "Using the aria-hidden="true" attribute on an element hides the element and all its children from screen readers and other assistive technologies. If the hidden element contains a focusable element, assistive technologies won't read the focusable element, but keyboard users will still be able to navigate to it, which can cause confusion."

We may be able to remove the `aria-hidden` from this if we have `visibility: hidden`, but for tabbing through the story, the `aria-hidden` doesn't disable the focusable elements (it does remove them from the a11y tree, but that's different)

**Comment by gmajoulet** [comment]
That's good to know thank you. So let's write the fix at the same level so the logic is not split in between two files.

The draggable drawer sets aria-hidden to true/false on open/close, and it also sets a `i-amphtml-story-draggable-drawer-open` class that you can rely on to set the visibility.
I suspect that setting the visibility on the Shadow host might not work and you could still tab through the content within the Shadow root. I won't get into specifics of shadow dom style scoping, but I think adding `:host {visibility: inherit !important;}` in both amp-story-bookend.css and amp-story-page-attachment.css will make it work.

**Comment by mszylkowski** [comment]
@gmajoulet had to deal with transitioning the visibility with a delay so it doesn't break the animations, but worked. Thanks for the suggestion! PTAL at the fix

File: extensions/amp-story/1.0/amp-story-share.css [hunk 2]
```diff
@@ -128,14 +128,9 @@
...
 
-.i-amphtml-story-share-icon:active,
 .i-amphtml-story-share-icon:focus {
-  outline: none !important;
-  box-shadow: none !important;
-}
-
-.i-amphtml-story-share-icon:active::before {
-  background-color: #787878 !important;
+  outline: 2px solid #0389ff !important;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-share.css:132]: Where does this color come from?

File: extensions/amp-story/1.0/amp-story-share.js [hunk 3]
```diff
@@ -103,6 +103,9 @@ const LINK_SHARE_ITEM_TEMPLATE = {
   tag: 'div',
   attrs: dict({
     'class': 'i-amphtml-story-share-icon i-amphtml-story-share-icon-link',
+    'tabindex': 0,
+    'role': 'button',
+    'aria-label': 'Get Link',
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-share.js:108]: i18n plz

File: extensions/amp-story/1.0/amp-story-share.css (hunk 2 above)
Comment (mszylkowski) [reply extensions/amp-story/1.0/amp-story-share.css:132]: https://amp.dev/documentation/components/amp-social-share/#indication-of-focus

File: extensions/amp-story/1.0/amp-story-bookend.css [hunk 4]
```diff
@@ -17,6 +17,10 @@
 @import './amp-story-shadow-reset.css';
 @import './amp-story-share.css';
 
+:host {
+  visibility: inherit !important;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-bookend.css:21]: It's worth adding an inline comment to say this propagates the visibility value from its shadow host for a11y

**Review by gmajoulet — APPROVED** [review]

File: extensions/amp-story/1.0/amp-story-share.css (hunk 2 above)
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-share.css:132]: nit:
It's best practice to not override default browser state for a11y. I think we can remove this override. 
The outline below it is a nice touch though to help it be visible.

It might be worth putting a comment above this block stating that it's for a11y.

File: extensions/amp-story/1.0/amp-story-share.js [hunk 5]
```diff
@@ -277,6 +289,13 @@ export class ShareWidget {
       e.preventDefault();
       this.copyUrlToClipboard_();
     });
+    listen(linkShareButton, 'keyup', (e) => {
+      const code = e.charCode || e.keyCode;
+      if (code === 32 || code === 13) {
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-share.js:294]: nit:
I wonder if this needs a comment for what keys these are for quick readability.
I think we commonly do top level CONSTs but maybe that's overkill for this.

**Review by processprocess — APPROVED** [review]
Approving with a couple nits/suggestions.

File: extensions/amp-story/1.0/amp-story-share.css (hunk 2 above)
Comment (mszylkowski) [reply extensions/amp-story/1.0/amp-story-share.css:132]: The default for these elements (being custom elements / divs and not necessarily buttons) is to not have anything for the focus state, so this is super useful to know what the user is focusing on. I can add a comment, but I wonder if it's not clear enough if it's inside a rule with the `:focus` selector

File: extensions/amp-story/1.0/amp-story-share.js (hunk 5 above)
Comment (mszylkowski) [reply extensions/amp-story/1.0/amp-story-share.js:294]: Sounds good to add a comment for this, it's not immediately obvious what these are
